src/bin/genemichaels.rs

```diff
@@ -250,14 +351,149 @@ fn main() {
                 Ok(_) => { },
                 Err(e) => {
                     if !args.quiet {
-                        eprintln!("Error formatting {}: {:?}", &file.to_string_lossy(), e);
+                        print_error_text();
+                        eprintln!("formatting {}: {:?}", &file.to_string_lossy(), e);
                     }
                     failed = true;
                 },
             };
         }
+        eprintln!(
+            "\x1B[1;32m    Finished\x1B[0;22m workspace formatted successfully in {:.2}s",
+            time::Instant::now().duration_since(inst).as_secs_f64()
+        );
         if failed {
             process::exit(1);
         }
     }
 }
+
+fn process_cargo_toml(path: PathBuf, manifest: Manifest, config: FormatConfig) {
+    let mut dirs = manifest.workspace.map(|wkspc| wkspc.members).map(|mb| {
+        let mut rust_dirs: Vec<PathBuf> =
+            mb.into_iter().filter_map(|m| path.join(&m).exists().then(|| path.join(m))).collect();
+        if let Some(bins) = manifest.bin {
```

Comment:
I tried with this `Cargo.toml` and it formatted `src/main.rs` but neither of the other source files:
```
[package]
name = "x103"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]

[[bin]]
name = "hello"
path = "hello/hello.rs"

[lib]
path = "mylib/lib.rs"
```

Reply:
I've made a couple changes, can you please retest this now? I'm still trying to understand exactly how it works but it seems this might fix our issue.

Reply:
Nice, thanks for the changes! I'll take a look.

Reply:
Ah, same issue.

I think there's two things going on here
1. All this happens within `manifest.workspace.map`, but it doesn't have anything to do with the workspace definition.  So if there's no workspace none of these bin/test/example things won't run.
2. The code currently says, "for each bin/test/example, if the bin/text/example's parent directory doesn't exist, add the default directory".  The adding the default directory should occur regardless of the existence/nonexistence of the directories - i.e. it shouldn't be in this for loop but rather down with the bit that adds the `"src"` directory.

Just as a note, regarding the `for` loops checking the parent directory:
- Before getting the parent directory I think you should join it to the Cargo.toml directory
- Parsing the path/getting the parent should never fail - if either fails then something's going wrong (maybe in the code) and we should either log that or abort and tell the user to fix their Cargo.toml file

Reply:
1. I think you're right, I've moved that code out, mapping the Option<Workspace> is definitely distinct to autobins etc, you're right.

2. Which `for` loop do you mean? The `for dir in dirs` or `for f in walkdir ..`, or another? I think I'm not quite understanding what problem this addresses, although either joining the Cargo.toml dir or running `canonicalize()` is probably a good idea, I'm just not sure where you need it.

Reply:
I agree this could be reported better, I've changed it to return an error instead. I think this is close to what you mean.

> FWIW I'm not sure why PathBuf::parse returns a Result but join doesn't... mysteries sweat_smile

I think it's because behind the scenes join just does `.push()`.